File: Raven.Tests/Faceted/ConditionalGetHelper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Text;
using Raven.Abstractions.Data;
using Raven.Client.Connection;
using Xunit;

namespace Raven.Tests.Faceted
{
    public static class ConditionalGetHelper
    {
        private static HttpWebResponse GetHttpResponseHandle304(WebRequest request)
        {
            try
            {
                return (HttpWebResponse)request.GetResponse();
            }
            catch (WebException e)
            {
                var response = e.Response as HttpWebResponse;
                if (response != null && response.StatusCode == HttpStatusCode.NotModified)
                {
                    return response;
                }

                throw;
            }
        }

		public static HttpStatusCode PerformGet(string url, Etag requestEtag, out Etag responseEtag)
        {
            var getRequest = WebRequest.Create(url);

            if(requestEtag != null)
                getRequest.Headers.Add("If-None-Match", requestEtag.ToString());

            using (var response = GetHttpResponseHandle304(getRequest))
            {
                try
                {
                    responseEtag = response.GetEtagHeader();
                }
                catch (Exception)
                {
                    responseEtag = null;
                }

                return response.StatusCode;
            }
        }

		public static HttpStatusCode PerformPost(string url, string payload, Etag requestEtag, out Etag responseEtag)
        {
            var request = WebRequest.Create(url);
            
            if (requestEtag != null)
                request.Headers.Add("If-None-Match", requestEtag.ToString());

            request.Method = "POST";

            byte[] buffer = Encoding.UTF8.GetBytes(payload);

            request.GetRequestStream().Write(buffer, 0, buffer.Length);

            using (var response = GetHttpResponseHandle304(request))
            {
                try
                {
                    responseEtag = response.GetEtagHeader();
                }
                catch (Exception)
                {
                    responseEtag = null;
                }

                return response.StatusCode;
            }
        }
    }
}

File: Raven.Client.Java/lib/src/test/java/raven/tests/faceted/ConditionalGetHelper.java
package raven.tests.faceted;

import java.io.IOException;

import org.apache.http.HttpResponse;
import org.apache.http.client.ClientProtocolException;
import org.apache.http.client.HttpClient;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.client.methods.HttpRequestBase;
import org.apache.http.entity.StringEntity;
import org.apache.http.impl.client.DefaultHttpClient;
import org.apache.http.util.EntityUtils;

import raven.abstractions.basic.Reference;
import raven.abstractions.data.Etag;
import raven.client.connection.HttpExtensions;

public class ConditionalGetHelper {
  private HttpClient httpClient;

  public ConditionalGetHelper() {
    httpClient = new DefaultHttpClient();
  }

  private HttpResponse getHttpResponseHandle304(HttpRequestBase request) throws ClientProtocolException, IOException {
    return httpClient.execute(request);
  }


  public int performGet(String url, Etag requestEtag, Reference<Etag> responseEtag) throws ClientProtocolException, IOException {
    HttpGet getRequest = new HttpGet(url);

    if (requestEtag != null) {
      getRequest.setHeader("If-None-Match", requestEtag.toString());
    }
    HttpResponse response = getHttpResponseHandle304(getRequest);

    try {
      responseEtag.value = HttpExtensions.getEtagHeader(response);
    } catch (Exception e) {
      responseEtag.value = null;
    }

    EntityUtils.consumeQuietly(response.getEntity());

    return response.getStatusLine().getStatusCode();
  }

  public int performPost(String url, String payload, Etag requestEtag, Reference<Etag> responseEtag) throws IOException {
    HttpPost request = new HttpPost(url);
    if (requestEtag != null) {
      request.setHeader("If-None-Match", requestEtag.toString());
    }

    request.setEntity(new StringEntity(payload));
    HttpResponse response = getHttpResponseHandle304(request);

    try {
      responseEtag.value = HttpExtensions.getEtagHeader(response);
    } catch (Exception e) {
      responseEtag.value = null;
    }

    EntityUtils.consumeQuietly(response.getEntity());

    return response.getStatusLine().getStatusCode();
  }
}
